### ISSUE TITLE
Dependencies instantiate at class definition time, breaking forward refs

### ISSUE THREAD

**Issue by eykd** [ZechCodes/Bevy#9]
Continuing on my quest to break bevy:

Dependencies for dataclasses appear to instantiate at class definition time, breaking any forward refs that might otherwise succeed under normal usage.

```python
from dataclasses import dataclass, field

from bevy import dependency


@dataclass
class Bar:
    baz: 'Baz' = field(default_factory=lambda: Baz())

@dataclass
class Foo:
    bar: Bar = dependency()


@dataclass
class Baz:
    value: str = 'baz'


def test_it_should_instantiate_foo_with_nested_forward_ref():
    foo = Foo()
    assert foo.bar.baz.value == "baz"
```

Test failure/traceback:

```
====================================================================================== ERRORS ======================================================================================
__________________________________________________________________________ ERROR collecting test_bevy.py ___________________________________________________________________________
.tox/py310/lib/python3.10/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
.tox/py310/lib/python3.10/site-packages/_pytest/runner.py:372: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
.tox/py310/lib/python3.10/site-packages/_pytest/python.py:531: in collect
    self._inject_setup_module_fixture()
.tox/py310/lib/python3.10/site-packages/_pytest/python.py:545: in _inject_setup_module_fixture
    self.obj, ("setUpModule", "setup_module")
.tox/py310/lib/python3.10/site-packages/_pytest/python.py:310: in obj
    self._obj = obj = self._getobj()
.tox/py310/lib/python3.10/site-packages/_pytest/python.py:528: in _getobj
    return self._importtestmodule()
.tox/py310/lib/python3.10/site-packages/_pytest/python.py:617: in _importtestmodule
    mod = import_path(self.path, mode=importmode, root=self.config.rootpath)
.tox/py310/lib/python3.10/site-packages/_pytest/pathlib.py:567: in import_path
    importlib.import_module(module_name)
/opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1050: in _gcd_import
    ???
<frozen importlib._bootstrap>:1027: in _find_and_load
    ???
<frozen importlib._bootstrap>:1006: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:688: in _load_unlocked
    ???
.tox/py310/lib/python3.10/site-packages/_pytest/assertion/rewrite.py:186: in exec_module
    exec(co, module.__dict__)
test_bevy.py:11: in <module>
    class Foo:
/opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/dataclasses.py:1184: in dataclass
    return wrap(cls)
/opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/dataclasses.py:1175: in wrap
    return _process_class(cls, init, repr, eq, order, unsafe_hash,
/opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/dataclasses.py:955: in _process_class
    cls_fields.append(_get_field(cls, name, type, kw_only))
/opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/dataclasses.py:730: in _get_field
    default = getattr(cls, a_name, MISSING)
.tox/py310/lib/python3.10/site-packages/bevy/injectors/classes.py:27: in __get__
    return repo.get(key)
.tox/py310/lib/python3.10/site-packages/bevy/repository.py:126: in get
    return self.create(key).value_or(default)
.tox/py310/lib/python3.10/site-packages/bevy/repository.py:80: in create
    match provider.create(key, repo):
.tox/py310/lib/python3.10/site-packages/bevy/providers/provider.py:35: in create
    cache[key] = factory()
<string>:3: in __init__
    ???
test_bevy.py:8: in <lambda>
    baz: 'Baz' = field(default_factory=lambda: Baz())
E   NameError: name 'Baz' is not defined
============================================================================= short test summary info ==============================================================================
ERROR test_bevy.py - NameError: name 'Baz' is not defined
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================================= 1 error in 0.11s =================================================================================
```

**Comment by ZechCodes** [comment]
`dependency` creates a descriptor that injects the correct value at access time. The dataclass is working at odds with that behavior. When the dataclass's class object is created and passed to the `dataclass` decorator it accesses each field with a default value. It then uses that value in `__init__` if no value was passed for that field. This not only causes the injection to happen prematurely it also breaks how descriptors work.

I can make this work by detecting when the dependency belongs to a dataclass and instead of returning the injected dependency return a field with a default factory that returns the injected dependency. This way if no value is passed to `__init__` for the field the dataclass will call a default factory that handles injection.

Here's a working solution that I can integrate into Bevy later this week:
```py
from dataclasses import dataclass, field

from bevy.injectors.classes import Dependency


class DataclassAwareDependency(Dependency):
    def __get__(self, instance, owner):
        if not instance and hasattr(owner, "__dataclass_params__"):
            getter = super().__get__
            return field(default_factory=lambda: getter(instance, owner))

        return super().__get__(instance, owner)


@dataclass
class Bar:
    baz: 'Baz' = field(default_factory=lambda: Baz())


@dataclass
class Foo:
    bar: Bar = DataclassAwareDependency()


@dataclass
class Baz:
    value: str = 'baz'


def test_it_should_instantiate_foo_with_nested_forward_ref():
    foo = Foo()
    assert foo.bar.baz.value == "baz"
```

**Comment by eykd** [comment]
Maybe `field_dependency` as a name?

**Comment by ZechCodes** [comment]
I think I'd just integrate it with the existing `Dependency` injector type. I think most people would expect the `dependency` function to work with dataclasses.

**Comment by ZechCodes** [comment]
be9566395b1466bd36828bda4b07ce1ffa8d57b2 resolves this issue and #7
